common/scheduler/schedule.c

```diff
@@ -591,14 +589,58 @@ sched_describetask(task_type* task)
     return strtask;
 }
 
-time_t
-sched_task_due(task_type* task)
+int
+schedule_task_istype(task_type* task, task_id type)
 {
-    return task->due_date;
+    return !strcmp(task->type, type);
 }
 
-int
-sched_task_istype(task_type* task, task_id type)
+void
+schedule_registertask(schedule_type* schedule, task_id taskclass, task_id tasktype, time_t (*callback)(task_type* task, char const *owner, void *userdata, void *context))
 {
-    return !strcmp(task->type, type);
+    struct schedule_handler* handlers;
+    handlers = realloc(schedule->handlers, sizeof(struct schedule_handler)*(schedule->nhandlers+1));
+    if (handlers != NULL) {
+        handlers[schedule->nhandlers].class    = taskclass;
+        handlers[schedule->nhandlers].type     = tasktype;
+        handlers[schedule->nhandlers].callback = callback;
+        schedule->handlers = handlers;
+        schedule->nhandlers += 1;
+    }
+}
+
+void
+schedule_scheduletask(schedule_type* schedule, task_id type, const char* owner, void* userdata, pthread_mutex_t* resource, time_t when)
+{
+    int i;
+    task_type* task;
+    struct schedule_handler* handler = NULL;
+    for (i = 0; i < schedule->nhandlers; i++) {
+        if (schedule->handlers[i].type == type) {
+            handler = &schedule->handlers[i];
+        }
+    }
+    if (handler) {
+        task = task_create(strdup(owner), handler->class, type, handler->callback, userdata, NULL, when);
+        task->lock = resource;
+        schedule_task(schedule, task, 0, 0);
+    }
+}
+
+void
+schedule_unscheduletask(schedule_type* schedule, task_id type, const char* owner)
```

Comment:
I know this is merged already but since it is so big I'd like to review it nonetheless. I think this function should take a class argument. Or alternatively be renamed to schedule_unschedule_signertask.